### PR TITLE
216_realise_rainfall_TaskRunner

### PR DESCRIPTION
**Pull request by Natali-natali**

### CODE REVIEW

File: src/main/java/org/codewars/utils/TaskRunner.java [hunk 1]
```diff
@@ -30,8 +32,14 @@ public static double runMean(){
         return 0;
     }
 
-    public static double runVariance(){
-        return 0;
+    public static void runVariance(){
+        System.out.println("Run task Rainfall");
```
Comment (mosvvy) [inline src/main/java/org/codewars/utils/TaskRunner.java:36]: Changes are needed:
or specify that it's variance part of task
or mean and variance should be merged in one run

Comment (Natali-natali) [reply src/main/java/org/codewars/utils/TaskRunner.java]: I meant to do it, but I forgot
done

**Review by RostislavKrotenko — DISMISSED** [review]

File: src/main/java/org/codewars/utils/TaskRunner.java [hunk 2]
```diff
@@ -30,8 +32,14 @@ public static double runMean(){
         return 0;
     }
 
-    public static double runVariance(){
-        return 0;
+    public void runVariance(){
+        System.out.println("Run task Rainfall: variance method");
+        System.out.println("Input the name of the city");
+        String town = Reader.readString();
+        System.out.println("Input string with rainfall records");
```
Comment (mosvvy) [inline src/main/java/org/codewars/utils/TaskRunner.java:39]: Mb rules should be specified here?
From Kata : "The records of towns are separated by \n. The name of each town is followed by :"

Comment (mosvvy) [reply src/main/java/org/codewars/utils/TaskRunner.java:39]: If we wouldn't show here task description, we should specify rules of such input - records of towns should be separated by '\n' and to separate town name and data use ':'

File: src/main/java/org/codewars/utils/TaskRunner.java [hunk 3]
```diff
@@ -30,8 +32,14 @@ public static double runMean(){
...
+        String str = Reader.readString();
+        double result = user.getSixImpl().variance(town, str);
+        System.out.println("The variance of rainfall for the city" + town + "is: " + result);
```
[the remark this answers is not in the document]
Comment (Natali-natali) [reply src/main/java/org/codewars/utils/TaskRunner.java:42]: What do you mean? First the city, then the result

[the remark this answers is not in the document]
Comment (Natali-natali) [reply src/main/java/org/codewars/utils/TaskRunner.java:42]: I got it

[the remark this answers is not in the document]
Comment (Natali-natali) [reply src/main/java/org/codewars/utils/TaskRunner.java:42]: about -1, seems it outputs the task

**Review by mosvvy — APPROVED** [review]

File: src/main/java/org/codewars/utils/TaskRunner.java [hunk 4]
```diff
@@ -39,7 +39,10 @@ public void runVariance(){
         System.out.println("Input string with rainfall records");
         String str = Reader.readString();
         double result = user.getSixImpl().variance(town, str);
-        System.out.println("The variance of rainfall for the city" + town + "is: " + result);
+        if(town != null){
+
+        }
```
Comment (mosvvy) [inline src/main/java/org/codewars/utils/TaskRunner.java:44]: empty if body

Comment (Natali-natali) [reply src/main/java/org/codewars/utils/TaskRunner.java:44]: I forgot to remove)

File: src/main/java/org/codewars/utils/TaskRunner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Natali-natali) [reply src/main/java/org/codewars/utils/TaskRunner.java]: In my opinion, this is not necessary

[the remark this answers is not in the document]
Comment (Natali-natali) [reply src/main/java/org/codewars/utils/TaskRunner.java]: I'm not sure, but I think such code will be hard to read

File: src/main/java/org/codewars/utils/TaskRunner.java [hunk 5]
```diff
@@ -1,108 +1,127 @@
...
 
-    public static double runVariance(){
+    public static long runFindNb() {
         return 0;
     }
 
-    public static String runBalance(){
+    public void rainfall() {
+        try {
+            System.out.println("Run task Rainfall: mean method");
```
Comment (mosvvy) [inline src/main/java/org/codewars/utils/TaskRunner.java:36]: "mean method" is not needed here

File: src/main/java/org/codewars/utils/TaskRunner.java [hunk 6]
```diff
@@ -1,108 +1,127 @@
...
+            System.out.println("Input the name of the city");
+            String town = Reader.readString();
+            System.out.println("Input string with rainfall records. City entries should be separated by \\n." +
+                    "The name of each city is followed by :");
+            String str = Reader.readString();
```
Comment (mosvvy) [inline src/main/java/org/codewars/utils/TaskRunner.java:41]: mb you've create branch from old dev? now Reader uses by reader (initializes in constructor)

**Comment by lhalam** [comment]
Conflicting files
src/main/java/org/codewars/utils/TaskRunner.java

File: src/main/java/org/codewars/utils/TaskRunner.java (hunk 6 above)
Comment (Natali-natali) [reply src/main/java/org/codewars/utils/TaskRunner.java:41]: If I'm not mistaken, Lubomyr recommended renaming the branches

File: src/main/java/org/codewars/utils/TaskRunner.java [hunk 7]
```diff
@@ -1,108 +1,127 @@
...
-        return 0.0;
-    }
-
-    public static long runFindNb(){
+    public static int runZeros() {
         return 0;
     }
 
-    public static double runMean(){
-        return 0;
```
Comment (lhalam) [inline src/main/java/org/codewars/utils/TaskRunner.java:30]: Do you really want to delete it?

Comment (Natali-natali) [reply src/main/java/org/codewars/utils/TaskRunner.java:30]: This is a task with two methods. So we combined them into one. What do you recommend, maybe we should create a new branchfor new method ?

**Comment by lhalam** [comment]
Conflicting files
src/main/java/org/codewars/utils/TaskRunner.java

File: src/main/java/org/codewars/utils/TaskRunner.java [hunk 8]
```diff
@@ -1,108 +1,127 @@
...
+            String str = Reader.readString();
+            double resultMean = user.getSixImpl().mean(town, str);
+            double resultVariance = user.getSixImpl().variance(town, str);
+            if (resultMean == -1 || resultVariance == -1) {
+                System.out.println("No such city in the records.");
+            } else {
+                System.out.println("The average of rainfall for the city of " + town + " is: " + resultMean);
+                System.out.println("The variance of rainfall for the city of " + town + " is: " + resultVariance);
+            }
+        } catch (InputMismatchException e) {
```
Comment (RostislavKrotenko) [inline src/main/java/org/codewars/utils/TaskRunner.java:50]: I'm not sure, but we have exceptions catching in Reader

**Pull request merged**